pymc/backends/base.py

```diff
@@ -254,14 +266,16 @@ def get_sampler_stats(self, stat_name, sampler_idx=None, burn=0, thin=1):
 
         if vals.dtype == np.dtype(object):
             try:
-                vals = np.vstack(vals)
+                vals = np.vstack(list(vals))
```

Comment:
I don't understand what happens here. So, after line 260 `vals` is an array of shape `(draws, len(sampler_idxs))`. At line 266 it is either `(draws, len(samplers_idx))` or `(draws,)`. Now, the `dtype` can be `object` iff `_get_sampler_stats` returns this kind of dtype, say, a `SamplerWarning`. So, after this line, the shape of `vals` is either still `(draws, len(samplers_idx))` or `(draws, 1)`. Did I miss something? So my questions are:
* In what case is the exception raised?
* Why do we remove the last singleton dimension to add it right back?
* What is the `list` doing?

Reply:
Yeah.. So this whole thing is a dirty hack to collapse object-arrays containing individual NumPy arrays to one array that maybe has a more informative dtype or shape.

For example:

first create an array containing object arrays:
```
>>> a = np.empty(2, dtype=object)
>>> a
array([None, None], dtype=object)
>>> a[0] = np.ones(2, dtype=object)
>>> a[1] = np.zeros(2, dtype=object)
>>> a
array([array([1, 1], dtype=object), array([0, 0], dtype=object)],
      dtype=object)
```
A `vstack` magically collapses them to one object array:
```
>>> np.vstack(a)
array([[1, 1],
       [0, 0]], dtype=object)
```

Unless they have different shapes, which triggers that `ValueError`.
```
>>> a[0] = np.ones(3, dtype=object)
>>> a
array([array([1, 1, 1], dtype=object), array([0, 0], dtype=object)],
      dtype=object)
>>> np.vstack(a)
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "<__array_function__ internals>", line 180, in vstack
  File "C:\Users\osthege\Miniconda3\envs\pm500\lib\site-packages\numpy\core\shape_base.py", line 282, in vstack
    return _nx.concatenate(arrs, 0)
  File "<__array_function__ internals>", line 180, in concatenate
ValueError: all the input array dimensions for the concatenation axis must match exactly, but along dimension 1, the array at index 0 has size 3 and the array at index 1 has size 2
>>>
```

The `list(vals)` is just to make mypy happy because `vstack(sequence)` but `isinstance(myndarray, Sequence) == False`.

----------

The whole construct is a massive PITA because it makes the shape of sampler stats very hard to predict.